packages/core/src/validate/run.ts

```diff
@@ -195,6 +224,71 @@ export function validate(solcOutput: SolcOutput, decodeSrc: SrcDecoder, solcVers
   return validation;
 }
 
+function checkNamespaceSolidityVersion(source: string, solcVersion?: string, solcInput?: SolcInput) {
+  if (solcVersion === undefined || solcInput === undefined) {
+    // These params should only be missing if using an old version of the Hardhat or Truffle plugin.
+    // Even without these params, namespace layout checks would still occur if compiled with solc version >= 0.8.20
+    debug('Cannot check Solidity version for namespaces because solcVersion or solcInput is undefined');
+  } else {
+    // Solc versions older than 0.8.20 do not have documentation for structs.
+    // Use a regex to check for strings that look like namespace annotations, and if found, check that the compiler version is >= 0.8.20
+    const content = solcInput.sources[source].content;
+    const hasMatch = content !== undefined && content.match(/@custom:storage-location\s+/);
+    if (hasMatch && versions.compare(solcVersion, '0.8.20', '<')) {
+      throw new UpgradesError(
+        `${source}: Namespace annotations require Solidity version >= 0.8.20, but ${solcVersion} was used`,
+        () =>
+          `Structs with the @custom:storage-location annotation can only be used with Solidity version 0.8.20 or higher. Use a newer version of Solidity, or remove the annotation if the struct is not used for namespaced storage.`,
+      );
+    }
+  }
+}
+
+function checkNamespacesOutsideContract(source: string, solcOutput: SolcOutput, decodeSrc: SrcDecoder) {
+  for (const node of solcOutput.sources[source].ast.nodes) {
+    if (isNodeType('StructDefinition', node)) {
+      const storageLocationArg = getStorageLocationArg(node);
+      if (storageLocationArg !== undefined) {
+        throw new UpgradesError(
+          `${decodeSrc(node)}: struct ${node.name} is defined outside of a contract`,
+          () =>
+            `Structs with the @custom:storage-location annotation must be defined within a contract. Move the struct definition into a contract, or remove the annotation if the struct is not used for namespaced storage.`,
+        );
+      }
+    }
+  }
+}
+
+function getNamespacedCompilationContext(
+  source: string,
+  contractDef: ContractDefinition,
+  namespacedOutput?: SolcOutput,
+) {
+  if (namespacedOutput === undefined || contractDef.canonicalName === undefined) {
```

Comment:
For `UserDefinedValueTypeDefinition`, canonical name includes the contract name e.g. `<contract>.<uvdt>`, but I'm not sure what's the difference (if any) for `ContractDefinition`

However, this code is reachable even if users are on < 0.8.9 and not using namespaces.  In this case, `canonicalName` is undefined for all contracts so we should effectively skip (by returning undefined as currently written).

Reply:
My guess is that canonicalName is the "scoped" name, but since contracts can't be nested they will always be defined at the root and `name` will be equal to `canonicalName`. It's nice not to encode the assumption of not being nested so it's good to use `canonicalName`.